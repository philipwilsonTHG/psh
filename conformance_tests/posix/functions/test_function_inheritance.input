# Test function inheritance in subshells
# PSH supports function inheritance in command substitution (v0.28.8 feature)

# Define test functions
parent_func() {
    echo "Parent function called with: $*"
}

math_func() {
    local a="$1"
    local b="$2"
    echo $((a * b))
}

recursive_func() {
    local n="$1"
    if [ "$n" -le 1 ]; then
        echo 1
    else
        echo $((n * $(recursive_func $((n - 1)))))
    fi
}

# Test function call in command substitution
echo "Testing function in command substitution:"
result=$(parent_func "hello" "world")
echo "Result: $result"

# Test function with arithmetic in command substitution
echo "Testing math function in command substitution:"
product=$(math_func 6 7)
echo "6 * 7 = $product"

# Test recursive function in command substitution
echo "Testing recursive function in command substitution:"
factorial_result=$(recursive_func 4)
echo "Factorial of 4 = $factorial_result"

# Test function in subshell with parentheses
echo "Testing function in subshell:"
(
    parent_func "subshell" "test"
    echo "Inside subshell, calling math_func:"
    math_func 3 4
)

# Test function inheritance in pipelines
echo "Testing function in pipeline:"
echo "5 8" | {
    read a b
    math_func "$a" "$b"
}

# Test nested command substitution with functions
nested_test_func() {
    local base="$1"
    echo "Base: $base, Double: $(math_func "$base" 2)"
}

echo "Testing nested command substitution:"
nested_result=$(nested_test_func 9)
echo "Nested result: $nested_result"

# Test function modification in subshell doesn't affect parent
modify_test() {
    echo "Original function"
}

echo "Before subshell:"
modify_test

(
    modify_test() {
        echo "Modified in subshell"
    }
    echo "Inside subshell:"
    modify_test
)

echo "After subshell:"
modify_test

# Test function with local variables in command substitution
local_test_func() {
    local temp="$1"
    local result=$((temp + 10))
    echo "Input: $temp, Result: $result"
}

echo "Testing function with locals in command substitution:"
local_result=$(local_test_func 25)
echo "Local test result: $local_result"