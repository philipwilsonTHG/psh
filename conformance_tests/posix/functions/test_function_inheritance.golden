Testing function in command substitution:
Result: Parent function called with: hello world
Testing math function in command substitution:
6 * 7 = 42
Testing recursive function in command substitution:
Factorial of 4 = 4
Testing function in subshell:
Parent function called with: subshell test
Inside subshell, calling math_func:
12
Testing function in pipeline:
0
Testing nested command substitution:
Nested result: Base: 9, Double: 0
Before subshell:
Original function
Inside subshell:
Modified in subshell
After subshell:
Modified in subshell
Testing function with locals in command substitution:
Local test result: Input: 25, Result: 35
